### PR TITLE
Add audio playback detection and screen sleep for audio only processes

### PR DESCRIPTION
**Pull request by g0r3**
This patch
* Adds automatic audio playback detection
* Allows caffeine to use powersaving for the screen if an audio only process is detected
* Keeps the screen on, if the the application playing back audio cannot specifically be identified as audio only (e.g. for videos).

### CODE REVIEW

File: caffeine/core.py [hunk 1]
```diff
@@ -120,16 +127,52 @@ def __attempt_autoactivation(self):
...
+        if not process_running and not fullscreen:
+            # Get all audio playback streams
+            audio_sinks = os.popen("pacmd list-sink-inputs").read()
+            for sink in audio_sinks.split("index:"):
+                is_running = False
+                role_music = False
+                for line in sink.split("\n"):
+                    if "state: RUNNING" in line:
+                        is_running = True
+                    if 'media.role = "music"' in line: # checked Spotify and Gnome Music. They use this attribute.
```
Comment (WhyNotHugo) [inline caffeine/core.py]: I don't see this role being used in any of the apps I have at hand.
However, `state: RUNNING` is only true if an app is actively playing back sound, so it might be enough to check just for that.

File: caffeine/core.py [hunk 2]
```diff
@@ -47,12 +47,15 @@ def __init__(self, process_manager):
...
-            XdgScreenSaverInhibitor(),
             XdgPowerManagmentInhibitor(),
             XssInhibitor(),
-            DpmsInhibitor(),
             XorgInhibitor(),
-            XautolockInhibitor()
+            XautolockInhibitor(),
+        ]
+
+        self.__screen_only_inhibitors = [
```
Comment (WhyNotHugo) [inline caffeine/core.py]: I'd rather add an attribute to each one `inhibits_suspension`, and use that (instead of keeping two lists). More on this below.

File: caffeine/core.py [hunk 3]
```diff
@@ -275,22 +318,39 @@ def __toggle_activated(self, note):
...
-    def _performTogglingActions(self):
+    def _performTogglingActions(self, inhibit_suspend, inhibit_screen):
         """This method performs the actions that affect the screensaver and
         powersaving."""
 
         for inhibitor in self.__inhibitors:
             if inhibitor.applicable:
-                logger.info("%s is applicable, running it.", inhibitor)
-                Thread(target=inhibitor.toggle).start()
+                logger.info("%s is applicable, setting it to %s." %(inhibitor, inhibit_suspend))
```
Comment (WhyNotHugo) [inline caffeine/core.py]: This whole set of changes in this `if` are a lot simpler if just using a `inhibits_suspension` attribute on each inhibitor.

File: caffeine/inhibitors.py [hunk 4]
```diff
@@ -71,7 +71,7 @@ def inhibit(self):
 
         self.__cookie = self.__proxy.Inhibit("Caffeine", dbus.UInt32(0),
                                              INHIBITION_REASON,
-                                             dbus.UInt32(12))
+                                             dbus.UInt32(4))
```
Comment (WhyNotHugo) [inline caffeine/inhibitors.py]: Why this change?

File: caffeine/core.py [hunk 5]
```diff
@@ -120,16 +127,52 @@ def __attempt_autoactivation(self):
...
                     logger.info("Fullscreen app detected. Inhibiting.")
 
-        if (process_running or fullscreen) and not self.__auto_activated:
+        # Let's look for playing audio:
+        self.music_procs = 0  # Number of supposed audio only streams. We can turn the screen off for those
+        screen_relevant_procs = 0  # Number of all audio streams including videos. We want to keep the screen on here
+
+        if not process_running and not fullscreen:
+            # Get all audio playback streams
+            audio_sinks = os.popen("pacmd list-sink-inputs").read()
```
Comment (WhyNotHugo) [inline caffeine/core.py]: I'm concerned about adding `pacmd` as an extra dependency.
Is there some way of reading this without it (eg: some PA socket? or a PA python library?).

File: caffeine/core.py (hunk 1 above)
Comment (g0r3) [reply caffeine/core.py]: I found this role to be used by Spotify and Gnome Music. So it seems to me this is a way some pure music players announce themselves to pulseaudio. I use this property only to check if i can turn off the screen. If this property is not present, Caffeine will still kick in, but also keep the screen on. Therefore I am checking for both properties.

File: caffeine/inhibitors.py (hunk 4 above)
Comment (g0r3) [reply caffeine/inhibitors.py]: This was necessary, so that GnomeInhibitor doesn't keep the screen on. This way it prevents suspension, but allows the screen to be turned off.

Please check the documentation for the flags here: https://people.gnome.org/~mccann/gnome-session/docs/gnome-session.html#org.gnome.SessionManager.Inhibit

File: caffeine/core.py (hunk 5 above)
Comment (g0r3) [reply caffeine/core.py]: > I'm concerned about adding `pacmd` as an extra dependency.
> Is there some way of reading this without it (eg: some PA socket? or a PA python library?).

At least on Arch Linux pacmd is part of pulseaudio itself. You cannot use pulseaudio without having this tool onboard too.

Comment (g0r3) [reply caffeine/core.py]: I don't use ALSA directly, but users who do will be locked out of this feature. Is there a way to do this with only ALSA? If so, all users will benefit. 

Afaik pulseaudio sits on top on ALSA, since ALSA can only handle one audio stream at a time and needs another tool to do the mixup of multiple streams into one. As far as I can tell almost no one uses ALSA without pulseaudio. And also: This pull request is intended to solve the issues for my use case (which should be congruent with the vast majority of users). ALSA only users don't gain anything here, that's right. But they also don't lose any functionality. So I don't intend to invest time to make this possible (anyone else is invited to do so though).

The thing with the pulsectl library is very much doable though. Thanks for the input

File: caffeine/core.py (hunk 3 above)
Comment (g0r3) [reply caffeine/core.py]: I will try that out. Thank you.

Comment (g0r3) [reply caffeine/core.py]: Ok, I propose following solution. I provided the Inhibitor base class with a `is_screen_inhibitor ` property, which gets overridden to return `True` for the inhibitors I had in the second list earlier:
```python
    def is_screen_inhibitor(self):
        return True
```

I also renamed the toggle method of the Inhibitor base class to `set`, which takes a bool. This makes it easier to control them individually, since the Inhibitors now can have a different state one to an other, and toggle would only invert this state:
```python
def set(self, state):
        if state:
            self.inhibit()
        else:
            self.uninhibit()
```

This allows to merge the lists back together and condenses the `_performTogglingActions` method to this:
```python
    def _performTogglingActions(self, suspend, susp_screen):
        for inhibitor in self.__inhibitors:
            if inhibitor.applicable:
                inhibitor.set(susp_screen) if inhibitor.is_screen_inhibitor() \
                                              else inhibitor.set(suspend)
                logger.info("%s is applicable, state: %s"
                            %(inhibitor, inhibitor.running))
        self.__inhibition_successful = not self.__inhibition_successful
```

File: caffeine/core.py (hunk 5 above)
Comment (g0r3) [reply caffeine/core.py]: > If you want to avoid dbus, you could use a library such as python-pulse-control. 

I was having a look on the library. While it works flawlessy when retrieving the audio sinks and sources, it doesn't seem to have an (easy) access the the `state: RUNNING` property of a source/sink, which seems to be provided by the `PA_SINK_INPUT_INFO` class in the pulsectl library. I need this in order to determine to deactivate suspending. The reason:

When you pause the audio playback, the audio sink/source (depending on the application) still remains in the list, but the state switches to `state: CORKED`. So in order to allow power suspension while paused audio playback applications are running, you need to have a look on this attribute.

So if no one can point me how to get this attribute via pulsectl, I would stick the solution which is implemented right now.

Comment (WhyNotHugo) [reply caffeine/core.py]: This looks okay to me -- I'm guessing `pactl` will have the same issues regarding Chrome, right? Since it probably just uses the same dbus interface.

Yeah, firefox has the same issue actually (and it keeps my headphones from powering down when inactive :cry: ).

I'd be happy to merge the PR using this interface -- it also looks much cleaner.

Comment (WhyNotHugo) [reply caffeine/core.py]: Also, this expects users to be using pulseaudio, if it turns out many people have an issue with this, I'll look into making it optional.

But for now, I'd avoid that effor, since it may not be an actual usecase anyway.

Comment (g0r3) [reply caffeine/core.py]: > Luckily for you, I spent hours figuring it all out with basically no documentation (I should help get that fixed if the dev wants more people to use it)

Haha, I thought the same when having a look. But thank you sitting through the pain. I've implemented it now and updated the pull request

File: caffeine/core.py (hunk 3 above)
Comment (g0r3) [reply caffeine/core.py]: I've updated the pull request with these changes. Please tell me what you think

Comment (g0r3) [reply caffeine/core.py]: And would it be ok for you to resolve the last couple of travis errors yourself? I don't really get how it is supposed to look and I don't want to spam commits just to solve this.
